README.md
# ⚠️⚠️⚠️ This repo just fork [react-native-svg-app-icon](https://github.com/aeirola/react-native-svg-app-icon) and remove icon shadow.

## Installation

```bash
yarn add -D react-native-svg-app-icon-without-shadow
```

## Usage

Place your square 108x108 SVG app icon file named `icon.svg` in the project root and run

```bash
npx react-native-svg-app-icon-without-shadow
```


----

# react-native-svg-app-icon

<p align="center">
  <img alt="Android adaptive icon"
    align="middle"
    src="src/__tests__/fixtures/example/icon.svg" />
  <img alt="iOS icon"
    align="middle"
    src="src/__tests__/fixtures/example/ios/example/Images.xcassets/AppIcon.appiconset/iphone-29@3x.png" />
  <img alt="Android legacy square icon"
    align="middle"
    src="src/__tests__/fixtures/example/android/app/src/main/res/mipmap-xhdpi/ic_launcher.png" />
  <img alt="Android legacy round icon"
    align="middle"
    src="src/__tests__/fixtures/example/android/app/src/main/res/mipmap-xhdpi/ic_launcher_round.png" />
</p>

![Build Status](https://github.com/aeirola/react-native-svg-app-icon/actions/workflows/main.yml/badge.svg?branch=master)

CLI tool for generating all the necessary iOS and Android application launcher icons for React Native projects from a single SVG source file. Features include:

- iOS PNG icon generation
- Android 8.0, and higher, vector drawable adaptive icon generation with PNG fallback
- Android 7.1 legacy circular icon generation
- Android 7.0, and lower, legacy square icon generation

For more information about the generated files, see the [docs](docs/generated_files.md).

:curly_loop: If you also want to use SVG images within your React Native application, you might want to check out [react-native-svg-asset-plugin](https://github.com/aeirola/react-native-svg-asset-plugin).

## Installation

```bash
npm install --save-dev react-native-svg-app-icon
```

SVG rendering handled by the splendid [`sharp`](https://github.com/lovell/sharp) library, meaning no dependencies outside of npm is required.

Requires node version 10, or later.

## Usage

Place your square 108x108 SVG app icon file named `icon.svg` in the project root and run

```bash
npx react-native-svg-app-icon
```

This will generate all the required icons under the `android/` and `ios/` directories.

### Icon background

If you want to use a separate background layer for Android adaptive icons, or because your source icon file doesn't contain a background, you can create an `icon-background.svg` file which will be used as the background layer for the generated icons.

In case you want to produce both foreground and background layers from a single SVG file, you can use [svg-deconstruct](https://github.com/not-fred/svg-deconstruct) to split layers to separate files. See configuration section below on how to specify input file paths.

## Configuration

Behaviour can be configured in the `app.json` under the `svgAppIcon` field. For example if you want to store icon layers under an `icon/` directory, you might want to use:

```json
{
  "name": "example",
  "displayName": "example",
  "svgAppIcon": {
    "foregroundPath": "./icon/icon-foreground.svg",
    "backgroundPath": "./icon/icon-background.svg",
    "platforms": ["ios"]
  }
}
```

Supported configuration values are

| Field            | Default                   | Description                                                                                                                                                     |
| ---------------- | ------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `foregroundPath` | `"./icon.svg"`            | Input file path for the foreground layer. File needs to exist, and may contain transparency.                                                                    |
| `backgroundPath` | `"./icon-background.svg"` | Input file path for the background layer. File doesn't need to exist, and will default to a fully white background. If file exist, it needs to be fully opaque. |
| `platforms`      | `["android", "ios"]`      | Array of platforms for which application launcher icons should be generated. Possible values are `android` and `ios`.                                           |

## Icon format

The input icon should be a SVG file adhering to the [Android adaptive icon specification](https://developer.android.com/guide/practices/ui_guidelines/icon_design_adaptive). Specifically, the image should:

- Be a valid SVG image
- have a 1:1 aspect ratio
- Have a size of 108x108dp

of which the:

- Center 72x72dp square is the normally visible area
- Center 66dp diameter circle is the safe area which will always be visible

With the various icons cropped according to the following image

![Icon copping anatomy](cropping.svg)

- ![#444](https://placehold.it/15/444?text=+) Overflow area
- ![#666](https://placehold.it/15/666?text=+) Visible area
- ![#888](https://placehold.it/15/888?text=+) iOS / Android legacy square crop
- ![#AAA](https://placehold.it/15/AAA?text=+) Android legacy circular crop
- ![#CCC](https://placehold.it/15/CCC?text=+) Safe area
- ![#F00](https://placehold.it/15/F00?text=+) Icon keylines

For an example icon file, see [`example/icon.svg`](example/icon.svg).

## Rationale

React Native aims to provide tools for building cross platform native mobile applications using technologies familiar from web development. Since the core tooling doesn't provide a solution for building the laundher icons for those applications, this tool aims to fill that gap.

Luckily, most icons follow a similar structure of a foreground shape on a background, which is easily adapted to different shapes and sizes. This is the idea behind Android [Adaptive Icons](https://developer.android.com/guide/practices/ui_guidelines/icon_design_adaptive), and what the [Android Image Asset Studio](https://developer.android.com/studio/write/image-asset-studio) implements nicely for generating leagy icons. This tool can actually be though of as a NPM CLI port of the Image Asset Studio, with added support for generating iOS icons as well.

### Other work

Most existing solutions are centered around the idea of scaling PNG images.

- [Expo](https://docs.expo.io/versions/latest/guides/app-icons/): Scales PNG files generating the required iOS and Android variants, but requires users to supply platform specific PNGs in order to adhere to platform icon design guidelines.
- [app-icon](https://github.com/dwmkerr/app-icon): Similar to Expo, with some added features such as labeling the icons. Requires imagemagick.

## Troubleshooting

### Supported SVG features

Most common SVG features are supported, including masks and styles. The underlying SVG rendering library is [`librsvg`](https://developer.gnome.org/rsvg/stable/rsvg.html) which claims to support most SVG 1.1 features, excluding scripts, animations and SVG fonts.

## Future improvements

- Add generation of Android notification icons


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.4.1](https://github.com/SolidZORO/react-native-svg-app-icon-without-shadow/compare/v0.4.0...v0.4.1) (2021-07-26)


### Bug Fixes

* fixed bin name ([078fcca](https://github.com/SolidZORO/react-native-svg-app-icon-without-shadow/commit/078fcca4f14ddddd1b33eb9e369c72c013d3835a))


src/android.ts
import * as path from "path";
import svg2vectordrawable from "svg2vectordrawable";

import * as input from "./input";
import * as output from "./output";

const adaptiveIconMinSdk = 26;

const densities: { name: ResourceDensity; scale: number }[] = [
  { name: "mdpi", scale: 1 },
  { name: "hdpi", scale: 1.5 },
  { name: "xhdpi", scale: 2 },
  { name: "xxhdpi", scale: 3 },
  { name: "xxxhdpi", scale: 4 }
];

const launcherName = "ic_launcher";
const roundIconName = "ic_launcher_round";
const launcherBackgroundName = "ic_launcher_background";
const launcherForegroundName = "ic_launcher_foreground";

/** Adaptive Icon **/
const adaptiveIconBaseSize = 108;
const adaptiveIconContent = (
  launcherBackgroundType: ResourceType,
  launcherForegroundType: ResourceType
): string => `<?xml version="1.0" encoding="utf-8"?>
<adaptive-icon xmlns:android="http://schemas.android.com/apk/res/android">
    <background android:drawable="@${launcherBackgroundType}/${launcherBackgroundName}" />
    <foreground android:drawable="@${launcherForegroundType}/${launcherForegroundName}" />
</adaptive-icon>`;

/** Legacy Icon **/

const legacyIconBaseSize = 48;
const inputIconContentRatio = input.inputContentSize / input.inputImageSize;

// Based on images from image asset studio at
// https://android.googlesource.com/platform/tools/adt/idea/+/refs/heads/mirror-goog-studio-master-dev/android/resources/images/launcher_stencil/
// https://android.googlesource.com/platform/tools/adt/idea/+/refs/heads/mirror-goog-studio-master-dev/android/src/com/android/tools/idea/npw/assetstudio/LauncherLegacyIconGenerator.java
const legacyLightningFilter = `
  <filter id="legacyLightningFilter">
    <!-- Drop shadow -->
    <feGaussianBlur in="SourceAlpha" stdDeviation="0.4" />
    <feOffset dx="0" dy="1.125" />
    <feComponentTransfer>
      <feFuncA type="linear" slope="0.2"/>
    </feComponentTransfer>
    <feComposite in2="SourceAlpha" operator="out"
      result="shadow"
    />

    <!-- Edge shade -->
    <feComponentTransfer in="SourceAlpha" result="opaque-alpha">
      <feFuncA type="linear" slope="0.2"/>
    </feComponentTransfer>
    <feOffset dx="-0.2" dy="-0.2" in="SourceAlpha" result="offset-alpha" />
    <feComposite in="opaque-alpha" in2="offset-alpha" operator="out"
      result="edge"
    />

    <feMerge>
<!--      <feMergeNode in="shadow" />-->
<!--      <feMergeNode in="edge" />-->
    </feMerge>
  </filter>`;

/** Legacy Square Icon **/
const legacySquareIconContentSize = 38;
const legacySquareIconBorderRadius = 3;
const legacySquareIconMargin =
  (legacyIconBaseSize - legacySquareIconContentSize) / 2;
const legacySquareIconContentRatio =
  legacySquareIconContentSize / legacyIconBaseSize;

const legacySquareIconMask = Buffer.from(
  `<svg version="1.1" xmlns="http://www.w3.org/2000/svg"
  viewBox="${getViewBox(legacySquareIconContentSize)}"
  width="${input.inputImageSize}" height="${input.inputImageSize}">
    <rect
      x="${legacySquareIconMargin}" y="${legacySquareIconMargin}"
      width="${legacySquareIconContentSize}" height="${legacySquareIconContentSize}"
      rx="${legacySquareIconBorderRadius}" ry="${legacySquareIconBorderRadius}"
    />
</svg>`,
  "utf-8"
);

const legacySquareIconOverlay = Buffer.from(
  `<svg version="1.1" xmlns="http://www.w3.org/2000/svg"
  viewBox="${getViewBox(legacySquareIconContentSize)}"
  width="${input.inputImageSize}" height="${input.inputImageSize}">
    ${legacyLightningFilter}
    <rect
      x="${legacySquareIconMargin}" y="${legacySquareIconMargin}"
      width="${legacySquareIconContentSize}" height="${legacySquareIconContentSize}"
      rx="${legacySquareIconBorderRadius}" ry="${legacySquareIconBorderRadius}"
      filter="url(#legacyLightningFilter)"
    />
</svg>`,
  "utf-8"
);

/** Legacy Round Icon **/
const legacyRoundIconContentSize = 44;
const legacyRoundIconContentRatio =
  legacyRoundIconContentSize / legacyIconBaseSize;

const roundIconMask = Buffer.from(
  `<svg version="1.1" xmlns="http://www.w3.org/2000/svg"
  viewBox="${getViewBox(legacyRoundIconContentSize)}"
  width="${input.inputImageSize}" height="${input.inputImageSize}">
    <circle
      cx="${legacyIconBaseSize / 2}" cy="${legacyIconBaseSize / 2}"
      r="${legacyRoundIconContentSize / 2}"
    />
</svg>`,
  "utf-8"
);

const roundIconOverlay = Buffer.from(
  `<svg version="1.1" xmlns="http://www.w3.org/2000/svg"
  viewBox="${getViewBox(legacyRoundIconContentSize)}"
  width="${input.inputImageSize}" height="${input.inputImageSize}">
    ${legacyLightningFilter}
    <circle
      cx="${legacyIconBaseSize / 2}" cy="${legacyIconBaseSize / 2}"
      r="${legacyRoundIconContentSize / 2}"
      filter="url(#legacyLightningFilter)"
    />
</svg>`,
  "utf-8"
);

function getViewBox(input: number): string {
  const size = input / inputIconContentRatio;
  const margin = (size - legacyIconBaseSize) / 2;
  const viewBox = [-margin, -margin, size, size];
  return viewBox.join(" ");
}

export interface Config {
  resDirPath: string;
  vectorDrawables: boolean;
}

export async function* generate(
  config: Partial<Config>,
  fileInput: input.FileInput
): AsyncIterable<string> {
  const fullConfig = getConfig(config);
  yield* generateLegacyIcons(fileInput, fullConfig);
  yield* generateRoundIcons(fileInput, fullConfig);
  yield* generateAdaptiveIcon(fileInput, fullConfig);
}

function getConfig(config: Partial<Config>): Config {
  return {
    resDirPath: config.resDirPath || "./android/app/src/main/res",
    vectorDrawables:
      config.vectorDrawables === undefined ? true : config.vectorDrawables
  };
}

async function* generateLegacyIcons(
  fileInput: input.FileInput,
  config: Config
): AsyncIterable<string> {
  yield* output.genaratePngs(
    {
      ...input.mapInput(fileInput, (inputData) => ({
        baseImage: inputData.backgroundImageData,
        operations: [
          { type: "composite", file: inputData.foregroundImageData.data },
          { type: "composite", blend: "mask", file: legacySquareIconMask },
          { type: "composite", file: legacySquareIconOverlay }
        ]
      })),
      cropSize: input.inputContentSize / legacySquareIconContentRatio
    },
    densities.map((density) => ({
      filePath: getIconPath(
        config,
        "mipmap",
        { density: density.name },
        `${launcherName}.png`
      ),
      outputSize: legacyIconBaseSize * density.scale
    }))
  );
}

async function* generateRoundIcons(
  fileInput: input.FileInput,
  config: Config
): AsyncIterable<string> {
  yield* output.genaratePngs(
    {
      ...input.mapInput(fileInput, (inputData) => ({
        baseImage: inputData.backgroundImageData,
        operations: [
          { type: "composite", file: inputData.foregroundImageData.data },
          { type: "composite", blend: "mask", file: roundIconMask },
          { type: "composite", file: roundIconOverlay }
        ]
      })),
      cropSize: input.inputContentSize / legacyRoundIconContentRatio
    },
    densities.map((density) => ({
      filePath: getIconPath(
        config,
        "mipmap",
        { density: density.name },
        `${roundIconName}.png`
      ),
      outputSize: legacyIconBaseSize * density.scale
    }))
  );
}

async function* generateAdaptiveIcon(
  fileInput: input.FileInput,
  config: Config
): AsyncIterable<string> {
  const backgroundImageInput = input.mapInput(fileInput, (inputData) => ({
    image: inputData.backgroundImageData
  }));
  let backgroundResourceType: ResourceType;
  try {
    yield* generateAdaptiveIconLayerVd(
      backgroundImageInput,
      launcherBackgroundName,
      config
    );
    backgroundResourceType = "drawable";
  } catch {
    yield* generateAdaptiveIconLayerPng(
      backgroundImageInput,
      launcherBackgroundName,
      config
    );
    backgroundResourceType = "mipmap";
  }

  const foregroundImageInput = input.mapInput(fileInput, (inputData) => ({
    image: inputData.foregroundImageData
  }));
  let foregroundResourceType: ResourceType;
  try {
    yield* generateAdaptiveIconLayerVd(
      foregroundImageInput,
      launcherForegroundName,
      config
    );
    foregroundResourceType = "drawable";
  } catch {
    yield* generateAdaptiveIconLayerPng(
      foregroundImageInput,
      launcherForegroundName,
      config
    );
    foregroundResourceType = "mipmap";
  }

  // Adaptive icon
  const adaptiveIconXml = adaptiveIconContent(
    backgroundResourceType,
    foregroundResourceType
  );
  yield* output.ensureFileContents(
    getIconPath(
      config,
      "mipmap",
      { density: "anydpi", minApiLevel: 26 },
      `${launcherName}.xml`
    ),
    adaptiveIconXml
  );
  yield* output.ensureFileContents(
    getIconPath(
      config,
      "mipmap",
      { density: "anydpi", minApiLevel: 26 },
      `${roundIconName}.xml`
    ),
    adaptiveIconXml
  );
}

async function* generateAdaptiveIconLayerVd(
  imageInput: input.Input<{ image: input.ImageData }>,
  fileName: string,
  config: Config
): AsyncIterable<string> {
  if (!config.vectorDrawables) {
    throw Error("Vector drawables disabled");
  }

  const imageData = await imageInput.read();
  const vdData = await svg2vectordrawable(
    imageData.image.data.toString("utf-8"),
    undefined,
    true
  );
  yield* output.ensureFileContents(
    getIconPath(
      config,
      "drawable",
      { density: "anydpi", minApiLevel: 26 },
      `${fileName}.xml`
    ),
    vdData
  );
}

async function* generateAdaptiveIconLayerPng(
  imageInput: input.Input<{ image: input.ImageData }>,
  fileName: string,
  config: Config
): AsyncIterable<string> {
  yield* output.genaratePngs(
    input.mapInput(imageInput, (imageData) => ({
      baseImage: imageData.image
    })),
    densities.map((density) => ({
      filePath: getIconPath(
        config,
        "mipmap",
        { density: density.name, minApiLevel: adaptiveIconMinSdk },
        `${fileName}.png`
      ),
      outputSize: adaptiveIconBaseSize * density.scale
    }))
  );
}

type ResourceType = "mipmap" | "drawable";

type ResourceDensity =
  | "ldpi"
  | "mdpi"
  | "hdpi"
  | "xhdpi"
  | "xxhdpi"
  | "xxxhdpi"
  | "anydpi";

function getIconPath(
  config: Config,
  resourceType: ResourceType,
  qualifier: {
    density: ResourceDensity;
    minApiLevel?: number;
  },
  fileName: string
): string {
  let directoryName: string[] = [resourceType];
  if (qualifier.density) {
    directoryName = [...directoryName, qualifier.density];
  }

  if (qualifier.minApiLevel) {
    directoryName = [...directoryName, `v${qualifier.minApiLevel}`];
  }
  return path.join(config.resDirPath, directoryName.join("-"), fileName);
}
